### PR TITLE
feat: revamp pricing and cost

### PR DESCRIPTION
**Pull request by maxdeichmann**

### CODE REVIEW

**Comment by maxdeichmann** [comment]
python sdk: https://github.com/langfuse/langfuse-python/pull/302

File: prisma/seed.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ const options = {
 
 const prisma = new PrismaClient();
 
-const TRACE_VOLUME = 100;
+const TRACE_VOLUME = 10_000;
```
Comment (maxdeichmann) [inline prisma/seed.ts:15]: revert

**Review by marcklingen — COMMENTED** [review]
wip, will also read latest changes

File: src/components/trace/index.tsx [hunk 2]
```diff
@@ -136,9 +136,11 @@ export function TracePage({ traceId }: { traceId: string }) {
 
   const totalCost: Decimal | undefined = trace.data?.observations.reduce(
     (prev: Decimal | undefined, curr: ObservationReturnType) => {
-      if (!curr.price) return prev;
-
-      return prev ? prev.plus(curr.price) : curr.price;
+      if (!curr.calculatedTotalCost) return prev;
+      console.log(curr.calculatedTotalCost);
```
Comment (marcklingen) [inline src/components/trace/index.tsx:140]: drop console

File: src/features/dashboard/components/LatencyChart.tsx [hunk 3]
```diff
@@ -36,7 +36,7 @@ export const LatencyChart = ({
         { column: "duration", agg: "90thPercentile" },
         { column: "duration", agg: "95thPercentile" },
         { column: "duration", agg: "99thPercentile" },
-        { column: "model" },
+        { column: "internalModel" },
```
Comment (marcklingen) [inline src/features/dashboard/components/LatencyChart.tsx:39]: discuss, should also show models here which do not have a price matched

File: src/features/dashboard/components/MetricTable.tsx [hunk 4]
```diff
@@ -26,13 +26,15 @@ export const MetricTable = ({
...
-        { column: "totalTokenCost" },
+        { column: "calculatedTotalCost", agg: "SUM" },
         { column: "totalTokens", agg: "SUM" },
-        { column: "model" },
+        { column: "internalModel" },
       ],
       filter: localFilters,
-      groupBy: [{ type: "string", column: "model" }],
-      orderBy: [{ column: "totalTokenCost", direction: "DESC" }],
+      groupBy: [{ type: "string", column: "internalModel" }],
```
Comment (marcklingen) [inline src/features/dashboard/components/MetricTable.tsx:34]: discuss, should also show models here which do not have a price matched

File: src/features/dashboard/components/ModelUsageChart.tsx [hunk 5]
```diff
@@ -36,8 +36,8 @@ export const ModelUsageChart = ({
       from: "observations",
       select: [
         { column: "totalTokens", agg: "SUM" },
-        { column: "totalTokenCost" },
-        { column: "model" },
+        { column: "calculatedTotalCost", agg: "SUM" },
+        { column: "internalModel" },
```
Comment (marcklingen) [inline src/features/dashboard/components/ModelUsageChart.tsx:40]: discuss, should also show models here which do not have a price matched

File: src/features/dashboard/components/hooks.ts [hunk 6]
```diff
@@ -11,7 +11,7 @@ export const getAllModels = (
     {
       projectId,
       from: "observations",
-      select: [{ column: "model" }],
+      select: [{ column: "internalModel" }],
```
Comment (marcklingen) [inline src/features/dashboard/components/hooks.ts:14]: discuss, should also show models here which do not have a price matched

**Review by marcklingen — COMMENTED** [review]
2/2

File: scripts/model-match.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+async function main() {
+  return await modelMatch();
+}
+
+// Call the function
+main().catch((err) => {
+  console.error("An error occurred:", err);
+});
+
+export async function modelMatch() {
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply scripts/model-match.ts:16]: add docstring

File: src/server/api/routers/generations.ts [hunk 8]
```diff
@@ -219,7 +202,7 @@ export const generationsRouter = createTRPCRouter({
             o.completion_tokens as "completionTokens",
             o.total_tokens as "totalTokens",
             o.version
-          FROM observations o
+          FROM observations_view o
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply src/server/api/routers/generations.ts:205]: add all types

File: src/server/api/services/tableDefinitions.ts [hunk 9]
```diff
@@ -77,51 +77,20 @@ export const totalTokens = {
   internal: 'o."total_tokens"',
 } as const;
 export const model = {
-  name: "model",
+  name: "internalModel",
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply src/server/api/services/tableDefinitions.ts:80]: revert to model for all tables and dashboards.

File: src/server/api/services/EventProcessor.ts [hunk 10]
```diff
@@ -34,6 +35,75 @@ export interface EventProcessor {
...
+  // or match pattern on the user provided model name
+  const modelCondition = existingObservation?.internalModel
+    ? Prisma.sql`AND model_name = ${existingObservation.internalModel}`
+    : model
+      ? Prisma.sql`AND ${model} ~ match_pattern`
+      : undefined;
+
+  // usage either from existing generation or from the current event
+  const mergedUnit = existingObservation?.unit ?? unit;
+
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply src/server/api/services/EventProcessor.ts:63]: other way round for unit and model merging

File: src/features/ingest/lib/usage.ts [hunk 11]
```diff
@@ -141,72 +143,15 @@ const getTokensByModel = (model: TiktokenModel, text: string) => {
...
-      "gpt-3.5",
-      "gpt-3.5-turbo",
-      "gpt-3.5-turbo-0301",
-      "gpt-3.5-turbo-0613",
-      "gpt-3.5-turbo-16k",
-      "gpt-3.5-turbo-16k-0613",
-      "gpt-3.5-turbo-1106",
-    ].find((m) => m === model) !== undefined
-  );
+  return !!model.includes(model as TiktokenModel);
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply src/features/ingest/lib/usage.ts:154]: delete

File: src/features/ingest/lib/usage.ts [hunk 12]
```diff
@@ -18,117 +59,78 @@ type ChatMessage = {
...
+  let numTokens = 0;
   params.messages.forEach((message) => {
-    num_tokens += tokens_per_message;
+    numTokens += params.config.tokensPerMessage;
 
     Object.keys(message).forEach((key) => {
       const value = message[key as keyof typeof message];
       if (value) {
-        num_tokens += getTokensByModel(params.model, value);
+        numTokens += getTokensByModel(params.model, value);
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply src/features/ingest/lib/usage.ts:104]: good, not this PR

File: scripts/model-match.ts [hunk 13]
```diff
@@ -16,79 +16,103 @@ main().catch((err) => {
...
 
-  const updatePromises = [];
-
-  for (const [key, observationsGroup] of Object.entries(groupedObservations)) {
-    // Split the key into its components
-    const [date, model, unit, projectId] = key.split("_");
+  // while observations are not null, get the next 10000 observations
+  let continueLoop = true;
+
+  while (continueLoop) {
```
Comment (maxdeichmann) [inline scripts/model-match.ts:23]: exit the loop if nothing was updated anymore

**Pull request merged**